File: lib/decisionEngine/decision.js
/**
 * decision.js — Decides WHETHER to reroute and logs the reasoning.
 *
 * ── Core problem this solves ─────────────────────────────────────────────────
 *
 * Without careful decision logic, a scoring function alone causes two bugs:
 *
 *   1. Oscillation: Route A → Route B → Route A → Route B (every 30 seconds)
 *      Fixed by: cooldown + stability score
 *
 *   2. Hair-trigger rerouting: Rerouting for a 10-second improvement
 *      Fixed by: minimum threshold gates before score is even checked
 *
 * ── Decision flow ────────────────────────────────────────────────────────────
 *
 *   shouldReroute() checks gates in order. First gate to fail stops the chain.
 *
 *   Gate 1 — Cooldown:        Has enough time passed since last reroute?
 *   Gate 2 — Same route:      Is the "best" route actually different?
 *   Gate 3 — Time saving:     Does candidate save >= minTimeSavingSeconds?
 *   Gate 4 — Traffic degraded: Has current route's congestion significantly worsened?
 *   Gate 5 — Score improvement: Does the composite score improve by enough?
 *
 *   Any single gate passing triggers a reroute.
 *   All gates failing returns { reroute: false }.
 *
 * ── State requirement ────────────────────────────────────────────────────────
 *
 * shouldReroute() needs a `currentRoute` object — the route currently in use.
 * This must be persisted externally (Redis, DB, or in-memory Map keyed by sessionId).
 * The caller manages persistence; this module manages logic.
 *
 * currentRoute must include:
 *   duration         — ETA in seconds when originally assigned
 *   congestion       — Congestion level at time of assignment
 *   baselineCongestion — Congestion level when the route was FIRST created
 *                        (used to detect degradation relative to original state)
 *   routeIndex       — Which Mapbox alternative was chosen
 *   geometry         — GeoJSON LineString (for stability comparison)
 *   assignedAt       — Unix ms timestamp when this route was set
 */

'use strict';

const { CITY_PROFILES, isPeakHour } = require('./config');

// ── Core decision function ────────────────────────────────────────────────────

/**
 * Determines whether the driver should switch to a new route.
 *
 * @param {Object} currentRoute   - The route currently active (see JSDoc above)
 * @param {Object} candidateRoute - Best-scored route from scoring.js
 * @param {Object} context        - { city, lastRerouteAt }
 *
 * @returns {{ reroute: boolean, reason: string, gate: string }}
 *   reason is a machine-readable slug suitable for logging and dashboards.
 *   gate indicates which decision gate triggered (or blocked) the reroute.
 */
function shouldReroute(currentRoute, candidateRoute, context = {}) {
  const {
    city          = 'default',
    lastRerouteAt = null,   // Unix ms, or null if never rerouted
  } = context;

  const profile    = CITY_PROFILES[city] ?? CITY_PROFILES.default;
  const thresholds = isPeakHour() ? profile.thresholds.peak : profile.thresholds.offPeak;
  const period     = isPeakHour() ? 'peak' : 'off_peak';

  // ── Gate 1: Cooldown ─────────────────────────────────────────────────────
  // Prevents oscillation. Even if a better route exists, we won't reroute
  // until cooldownMs has elapsed since the last reroute event.
  if (lastRerouteAt !== null) {
    const elapsed   = Date.now() - lastRerouteAt;
    const remaining = profile.cooldownMs - elapsed;
    if (remaining > 0) {
      return {
        reroute: false,
        gate:    'cooldown',
        reason:  `cooldown_active:${Math.ceil(remaining / 1000)}s_remaining`,
      };
    }
  }

  // ── Gate 2: Candidate is the same route ──────────────────────────────────
  // The scoring already chose the best route. If that's the current one,
  // there is nothing to switch to.
  if (candidateRoute.routeIndex === currentRoute.routeIndex) {
    return {
      reroute: false,
      gate:    'same_route',
      reason:  'best_route_unchanged',
    };
  }

  const timeSavedSeconds = currentRoute.duration - candidateRoute.duration;

  // ── Gate 3: Candidate saves significant time ──────────────────────────────
  // Pure time-saving gate: if the new route is materially faster, reroute
  // regardless of congestion changes.
  if (timeSavedSeconds >= thresholds.minTimeSavingSeconds) {
    return {
      reroute: true,
      gate:    'time_saving',
      reason:  `time_saving:${timeSavedSeconds}s_saved:period=${period}`,
    };
  }

  // ── Gate 4: Current route has significantly degraded ─────────────────────
  // Compares current congestion against baseline (when route was assigned).
  // If our own sensor network sees a large spike, and the candidate is
  // materially clearer, reroute.
  const baselineCongestion = currentRoute.baselineCongestion ?? currentRoute.congestion;
  const congestionIncrease = baselineCongestion > 0
    ? (currentRoute.congestion - baselineCongestion) / baselineCongestion
    : 0;

  const congestionDelta = currentRoute.congestion - candidateRoute.congestion;

  if (
    congestionIncrease  >= thresholds.minTrafficIncreasePct &&
    congestionDelta     >= thresholds.minCongestionDelta
  ) {
    return {
      reroute: true,
      gate:    'traffic_degradation',
      reason:  [
        `current_route_degraded`,
        `congestion_up_${(congestionIncrease * 100).toFixed(0)}pct`,
        `candidate_clearer_by_${congestionDelta.toFixed(0)}pts`,
        `period=${period}`,
      ].join(':'),
    };
  }

  // ── Gate 5: Composite score improvement is large enough ──────────────────
  // Falls back to the full multi-dimensional score when the other gates
  // don't provide a clear signal. Only triggers if candidate is materially better.
  if (currentRoute.score !== undefined && candidateRoute.score !== undefined) {
    const improvement = currentRoute.score > 0
      ? (currentRoute.score - candidateRoute.score) / currentRoute.score
      : 0;

    if (improvement >= thresholds.minScoreImprovement) {
      return {
        reroute: true,
        gate:    'score_improvement',
        reason:  `score_improvement:${(improvement * 100).toFixed(1)}pct:period=${period}`,
      };
    }
  }

  // ── No gate passed ────────────────────────────────────────────────────────
  return {
    reroute: false,
    gate:    'no_trigger',
    reason:  [
      `no_significant_improvement`,
      `time_saved=${timeSavedSeconds}s`,
      `min_required=${thresholds.minTimeSavingSeconds}s`,
      `period=${period}`,
    ].join(':'),
  };
}

// ── Route selector ────────────────────────────────────────────────────────────

/**
 * Selects the single best route from a scored + sorted array.
 * Wraps it with a human-readable selection reason for debug output.
 *
 * @param {Object[]} scoredRoutes - from scoring.js, sorted ascending
 * @returns {{ route: Object, reason: string }}
 */
function selectBestRoute(scoredRoutes) {
  if (scoredRoutes.length === 0) {
    throw new Error('[decision] No routes to select from');
  }

  const best = scoredRoutes[0];

  // Build a human-readable explanation of why this route won
  const dominant = dominantFactor(best.breakdown, best.normalized);

  const reason = [
    `route_${best.routeIndex}_selected`,
    `score=${best.score.toFixed(3)}`,
    `dominant_factor=${dominant}`,
    `eta=${Math.round(best.duration)}s`,
    `congestion=${best.congestion.toFixed(0)}`,
    `weather_penalty=${best.weatherPenalty.toFixed(2)}`,
    `stability=${best.stability.toFixed(2)}`,
  ].join(' | ');

  return { route: best, reason };
}

/**
 * Returns the dimension name that contributed most to the composite score.
 * Used purely for human-readable debug output.
 */
function dominantFactor(breakdown) {
  return Object.entries(breakdown)
    .sort(([, a], [, b]) => b - a)[0][0];
}

module.exports = { shouldReroute, selectBestRoute };

File: backend/decision_engine/decision.py
"""
decision.py — Five-gate rule-based reroute arbiter.

Gates are evaluated in order. The FIRST gate that fires determines the outcome.
If no gate fires, the engine holds the current route.

Gate order (same as Node.js version — must stay in sync):
  1. cooldown          → HOLD  (protect driver from thrashing)
  2. same_route        → HOLD  (no point switching to current route)
  3. time_saving       → REROUTE (candidate saves ≥ N seconds)
  4. traffic_degradation → REROUTE (current route got significantly worse)
  5. score_improvement → REROUTE (composite score meaningfully better)
  6. hold              → HOLD  (default — improvement too marginal)
"""
from __future__ import annotations

import logging
from datetime import datetime, timezone
from typing import Optional

from .config import CITY_PROFILES, classify_time_of_day
from .schemas import CurrentRouteState, RuleDecision, ScoredRoute

log = logging.getLogger(__name__)


def _profile(city: str) -> dict:
    return CITY_PROFILES.get(city, CITY_PROFILES["default"])


def _thresholds(city: str) -> dict:
    hour    = datetime.now(timezone.utc).hour
    tod     = classify_time_of_day(hour)
    period  = "peak" if tod == "peak" else "off_peak"
    return _profile(city)["thresholds"][period]


def should_reroute(
    current_route:   CurrentRouteState,
    candidate:       ScoredRoute,
    city:            str = "default",
    last_reroute_at: Optional[datetime] = None,
) -> RuleDecision:
    """
    Returns a RuleDecision(reroute, gate, reason).
    Never raises — any exception results in a HOLD with reason 'error'.
    """
    try:
        profile    = _profile(city)
        thresholds = _thresholds(city)
        cooldown_s = profile["cooldown_s"]

        # ── Gate 1: Cooldown ──────────────────────────────────────────────────
        if last_reroute_at is not None:
            now     = datetime.now(timezone.utc)
            elapsed = (now - last_reroute_at).total_seconds()
            if elapsed < cooldown_s:
                remaining = int(cooldown_s - elapsed)
                return RuleDecision(
                    reroute = False,
                    gate    = "cooldown",
                    reason  = f"cooldown_active:{remaining}s_remaining",
                )

        # ── Gate 2: Same route ────────────────────────────────────────────────
        if candidate.route_index == current_route.route_index:
            return RuleDecision(
                reroute = False,
                gate    = "same_route",
                reason  = "candidate_is_current_route",
            )

        # ── Gate 3: Time saving ───────────────────────────────────────────────
        time_saved = current_route.duration - candidate.duration
        if time_saved >= thresholds["min_time_saving_s"]:
            return RuleDecision(
                reroute = True,
                gate    = "time_saving",
                reason  = f"saves_{int(time_saved)}s_ge_{thresholds['min_time_saving_s']}s_threshold",
            )

        # ── Gate 4: Traffic degradation ───────────────────────────────────────
        if current_route.baseline_congestion > 0:
            congestion_increase_pct = (
                (current_route.congestion - current_route.baseline_congestion)
                / current_route.baseline_congestion
            )
            congestion_delta = current_route.congestion - candidate.congestion

            if (
                congestion_increase_pct >= thresholds["min_traffic_increase_pct"]
                and congestion_delta     >= thresholds["min_congestion_delta"]
            ):
                return RuleDecision(
                    reroute = True,
                    gate    = "traffic_degradation",
                    reason  = (
                        f"current_congestion_up_{congestion_increase_pct:.0%}_"
                        f"candidate_{congestion_delta:.0f}pts_clearer"
                    ),
                )

        # ── Gate 5: Score improvement ─────────────────────────────────────────
        if current_route.score > 0:
            improvement = (current_route.score - candidate.score) / current_route.score
            if improvement >= thresholds["min_score_improvement"]:
                return RuleDecision(
                    reroute = True,
                    gate    = "score_improvement",
                    reason  = f"score_improved_{improvement:.1%}",
                )

        return RuleDecision(
            reroute = False,
            gate    = "hold",
            reason  = "improvement_below_all_thresholds",
        )

    except Exception as exc:
        log.error("[decision] should_reroute raised: %s", exc, exc_info=True)
        return RuleDecision(reroute=False, gate="error", reason=str(exc))
